builtin/providers/openstack/resource_openstack_networking_secgroup_rule_v2.go

```diff
@@ -0,0 +1,195 @@
+package openstack
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/networking/v2/extensions/security/rules"
+)
+
+func resourceNetworkingSecGroupRuleV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNetworkingSecGroupRuleV2Create,
+		Read:   resourceNetworkingSecGroupRuleV2Read,
+		Delete: resourceNetworkingSecGroupRuleV2Delete,
+
+		Schema: map[string]*schema.Schema{
+			"region": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				DefaultFunc: schema.EnvDefaultFunc("OS_REGION_NAME", ""),
+			},
+			"direction": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ethertype": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"port_range_min": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"port_range_max": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"protocol": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
```

Comment:
When omitting `protocol`, but specifying a port, I get the following error:

```{"NeutronError": {"message": "Must also specify protocol if port range is given.", "type": "SecurityGroupProtocolRequiredWithPorts", "detail": ""}}```

In a way, this makes `protocol` semi-required.

It might be best to check for this condition in the `Create` function before passing it onto Gophercloud/Neutron. Something like (off the top of my head, may contain errors):

```
portRangeMin := d.Get("port_min_range").(string)
portRangeMax := d.Get("port_min_range").(string)
protocol := d.Get("protocol").(string)

if portRangeMin == "" && portRangeMax == "" && protocol == "" {
  return fmt.Errorf("A protocol must be specified when using port_range_min and port_range_max")
}
```